content/en/blog/2023/kwok/index.md

```diff
@@ -0,0 +1,96 @@
+
+## What is KWOK?
+
+KWOK stands for Kubernetes WithOut Kubelet. So far, it provides two tools:
+
+`kwok`
+: `kwok` is the cornerstone of this project, responsible for simulating the lifecycle of
+fake nodes, pods, and other Kubernetes API resources.
+
+`kwokctl`
+: `kwokctl` is a CLI tool designed to streamline the creation and management of clusters,
+with nodes simulated by `kwok`.
+
+## Why use KWOK?
+
+KWOK has several advantages:
+
+- **Lightweight**: You can simulate thousands of nodes on your laptop without significant consumption
+  of CPU or memory resources.
+- **Fast**: You can create and delete clusters and nodes almost instantly,
+  without waiting for boot or provisioning.
+- **Compatibility**: KWOK works with any tools or clients that are compliant with Kubernetes APIs,
+  such as kubectl, helm, etc.
+- **Portability**: KWOK has no specific hardware or software requirements.
+  You can run it using prebuilt images, once Docker/Podman/nerdctl is installed.
+  Alternatively, binaries are also available for all platforms and can be easily installed.
+- **Flexibility**: You can configure different node types, labels, taints, capacities, conditions, etc.,
+  and you can configure different pod behaviors, status, etc. to test different scenarios and edge cases.
+
+{{< image width="90%"
+    link="./manage-clusters.svg"
+    caption="Using kwokctl to manage simulated clusters"
+    >}}
+
+## Testing Istio with KWOK
+
+For the latest information, see [KWOK with Istio](https://kwok.sigs.k8s.io/docs/examples/istio/).
+
+To create a test environment for the Istio control plane with [KWOK](https://kwok.sigs.k8s.io/),
+you can follow these steps:
+
+### Set up Cluster
+
+{{< text bash >}}
+$ kwokctl create cluster --runtime kind
+{{< /text >}}
+
+### Create Node
+
+{{< text bash >}}
+$ kubectl apply -f https://kwok.sigs.k8s.io/examples/node.yaml
+{{< /text >}}
+
+### Deploy Istio
+
+{{< text bash >}}
+$ istioctl install -y
+{{< /text >}}
+
+### Migrate Controllers to Real Node
+
+We need to use [kind](https://kind.sigs.k8s.io/) to run the controller because KWOK itself cannot actually run a Pod.
+
+{{< text bash >}}
+$ kubectl patch deploy istiod -n istio-system --type=json -p='[{"op":"add","path":"/spec/template/spec/nodeName","value":"kwok-kwok-control-plane"}]'
+{{< /text >}}
+
+### Create Pod and Inject Sidecar
+
+{{< text bash >}}
+$ kubectl label namespace default istio-injection=enabled
+$ kubectl apply -f {{< github_file >}}/samples/bookinfo/platform/kube/bookinfo.yaml
+{{< /text >}}
+
+You will see that the Pod is running on a fake node. However, it is still an injected sidecar.
+
+## Limitations
+
+With KWOK alone, we can only simulate the cluster data,
+but the real environment is actually also connected with Sidecar's xDS, which cannot be simulated.
+So we need a KWOK based xDS client connection generator to complete the load simulation,
+I made a [PWOD](https://github.com/wzshiming/pwod) to simulate the xDS client connection.
```

Comment:
Saying you _made_ this feels a bit disingenuous. As far as I can tell, what was done is a library that *I made*, https://github.com/howardjohn/pilot-load/, was both directly copy+pasted and imported directly, without attribution. The whole library is 300 lines of code with virtually every line identically copied from `pilot-load`.

I am happy to see more work in performance testing area, and happy to collaborate in that regard. While I am sure you had good intentions this feels more like stealing my work and passing it off as your own (and then publicizing it on a blog that I am heavily involved in...).

Would it make sense to just directly use `pilot-load` here? As far as I can tell `pwod` is just a strict subset of `pilot-load` -- its not clear why another project is needed, even ignoring attribution issues.

Technically you could replace the entire blog with `pilot-load`, as it can do the work of `kwok` for the purpose of Istio testing... but I can understand why that would not be desired :slightly_smiling_face:

Reply:
![image](https://github.com/istio/istio.io/assets/132510/0719cb9e-c3fc-4fbc-849d-4d94e43c3527|width=250)

Reply:
> Saying you made this feels a bit disingenuous. As far as I can tell, what was done is a library that I made, https://github.com/howardjohn/pilot-load/, was both directly copy+pasted and imported directly, without attribution. The whole library is 300 lines of code with virtually every line identically copied from pilot-load.

The demo project is something I realized I missed while making this post.
The `go.mod` of the project in it references pilot-load directly and then rewrites the entry point, referencing only a few needed packages, for a quick implementation of my idea.

I [left a comment on your issue](https://github.com/howardjohn/pilot-load/issues/9#issuecomment-1641466182) almost the same day I made the demo to get your opinion.

> I am happy to see more work in performance testing area, and happy to collaborate in that regard. While I am sure you had good intentions this feels more like stealing my work and passing it off as your own (and then publicizing it on a blog that I am heavily involved in...).

I apologize for making you feel offended. 🙏🙏🙏

> Would it make sense to just directly use pilot-load here? As far as I can tell pwod is just a strict subset of pilot-load -- its not clear why another project is needed, even ignoring attribution issues.
> Technically you could replace the entire blog with pilot-load, as it can do the work of kwok for the purpose of Istio testing... but I can understand why that would not be desired 🙂

I agree with you.

3 years ago when I was actually doing a production press test for Istio, and I ended up abstracting two parts of the logic from the test, [wzshiming/xds](https://github.com/wzshiming/xds) and [fake-kubelet](https://github.com/wzshiming/fake-kubelet/tree/v0.2.0), originally I wanted to use the xds for the demo, but I felt that the dependency hadn't been maintained in 3 years, so I referenced the xds from pilot-load.

I took over this test task from someone else, that code that was hard-coded to inject test data into the Istio source code for the purpose of pressure testing, and I spent a lot of time learning about Istio and completely rewriting the test code.

Then I realized that most of the testers using it weren't interested in understanding the principles behind it and just wanted to test it in the most intuitive and simple way possible.

Reply:
I didn't mean to come off as super offended, mostly just surprised. I think kwok is a really great idea but it seems like the xds part can just be a part of pilot-load. I am happy to collaborate there, and possibly in this blog if you are interested, accept PRs there, even move it to a neutral location if desired. Let me know what interests you

Reply:
of course，thank you

Reply:
Could you add something that describes how to use the pilot-load with kwok?

Reply:
Just need to sue `pilot-load cluster`. The `--cluster-type` has 3 modes:

* real: actually create real pods. Can be combined with kwok to make those 0 cost
* fake-node: replace kwok basically. Create real pods but assign to fake nodes
* Fake: connect to a plain apiserver with no controller-manager at all